### ISSUE TITLE
Multinomial coefficients

### ISSUE THREAD

**Issue by sylee957** [sympy/sympy#18933]
I think that multinomial coefficient can be added
It won't be difficult, but the conflicting notations between mathematica and maple is making it confusing.

Some people prefer to use
- ![image](https://user-images.githubusercontent.com/34944973/77272609-f6673380-6cf4-11ea-911e-70d0526f4101.png)

while some other people prefer to use
- ![image](https://user-images.githubusercontent.com/34944973/77272641-08e16d00-6cf5-11ea-8233-583e4f2d9ce7.png)

assuming that n is loosely the summation of each k.

So it should be chosen that which kind if definition is the best
And if the second one is chosen, I think that I should rather make it like 
![image](https://user-images.githubusercontent.com/34944973/77273070-0895a180-6cf6-11ea-9ed6-e38119f1d571.png) mechanically, so it won't really be much convenient than choosing the first one. But this can be consistent with the definition of binomial.

Some references:
- https://en.wikipedia.org/wiki/Multinomial_theorem
- https://dlmf.nist.gov/26.4
- http://functions.wolfram.com/GammaBetaErf/Multinomial/
- https://www.maplesoft.com/support/help/Maple/view.aspx?path=combinat%2Fmultinomial
- https://reference.wolfram.com/language/ref/Multinomial.html

**Comment by Arpan612** [comment]
Sir, I would like to work on this issue. I will be sending a PR soon.

**Comment by sylee957** [comment]
Yes, but I'd expect it to be presented as a `class multinomial(Function):` rather than loose function. I think that there are already lots of same stuff implemented for computing mutinomial only for integers, but only not as a symbolic function

**Comment by Arpan612** [comment]
Sure sir.

**Comment by Arpan612** [comment]
Except for integers, where all should multinomial be applicable?

**Comment by sylee957** [comment]
In the wolfram function repository, I see they are using a generalized definition using gamma functions. So it can be defined all over the complexes and differentiable.

And I already see that factorial or binomial in sympy are using the definition from gamma, so the definition from wolfram can be used.

**Comment by Arpan612** [comment]
http://functions.wolfram.com/GammaBetaErf/Multinomial/02/
Sir, do you mean this one?

**Comment by sylee957** [comment]
Yes

**Comment by iammosespaulr** [comment]
@Arpan612 you haven't submitted any pr for this one either ... smh ...

**Comment by iammosespaulr** [comment]
@sylee957 would it be preferred to write the Multinomial from scratch or just implement it as a product of binomials ?

**Comment by Arpan612** [comment]
@iammosespaulr I am working on it. I will send a PR in a few hours. Please do not attempt it. There are more than 2900 unclaimed issues. Please work on it.

**Comment by iammosespaulr** [comment]
@Arpan612 I am sorry .... but this issue has been left open for 2 days ... just because you are working on it doesn't mean that I shouldn't ... if you complete it send a PR ... and If I complete it first I will do so ...

**Comment by sylee957** [comment]
> @sylee957 would it be preferred to write the Multinomial from scratch or just implement it as a product of binomials ?

In my opinion, I'd just implement it from scratch because expanding a multinomial into a product of binomials may not have a unique representation.
But the computation can be done internally by that.

**Comment by Arpan612** [comment]
@iammosespaulr Now please allow me to work on this. @sylee957 I request you to make it clear now that only I will be working on this. I will submit a final PR in a few hours.

**Comment by iammosespaulr** [comment]
> > @sylee957 would it be preferred to write the Multinomial from scratch or just implement it as a product of binomials ?
> 
> In my opinion, I'd just implement it from scratch because expanding a multinomial into a product of binomials may not have a unique representation.
> But the computation can be done internally by dispatching them to binomials.

beautiful ... okay then

**Comment by Arpan612** [comment]
@iammosespaulr You are intentionally trying to create a conflict by taking up claimed issues when you can work on unclaimed ones. I believe you have joined this organization today. Please do not treat your fellow contributors in this manner. @sylee957 @asmeurer @czgdp1807 Please look into his unruly behavior.

**Comment by iammosespaulr** [comment]
@Arpan612 I don't think that there is such thing as 'claiming issues' ... everyone is free to work on improving code and the code which the admins see fit, is merged. I don't have any conflict with you.

**Comment by Arpan612** [comment]
@iammosespaulr Working simultaneously on the same issue creates unnecessary tension, someone with prior experience of open source will know. And as @sylee957 said, the one who submits a PR first should have the full right to work on it. So, you will look for another issue now if you actually do not want conflicts.

**Comment by Arpan612** [comment]
@iammosespaulr If no other issues existed, working on the same issue makes sense. But as I said, more than 2900 issues are there which are unclaimed/abandoned for weeks. Please feel free to contribute there.

**Comment by iammosespaulr** [comment]
let's just agree to disagree.

**Comment by abhinav28071999** [comment]
@Arpan612 If i am not wrong then by your logic, #18950  should be preferred against #18951 cause @iammosespaulr submitted the pr first. ?????

**Comment by Arpan612** [comment]
@abhinav28071999 It is common courtesy to allow a person who has claimed an issue just "5 hours" ago to give at least a day. Even more, a reason to leave the issue is when the person immediately says he is working.

**Comment by abhinav28071999** [comment]
As far as i know there is no claiming of issues in Sympy. If you like an issue and if there is no ongoing PR fixing it then anyone can take it. You might say that you claimed an issue but who knows someone else might be already working on it and  didn't bother to send a claiming msg. So it's better that if anyone want to take up an issue then submit a pr for fixing it first otherwise you could help the other person with improvements. Moreover after claiming an issue who knows the person might be able to fix it or not. So it's much better to submit a reasonable pr.

**Comment by czgdp1807** [comment]
Please try to avoid off topic discussions. Some comments here were not relevant to the issue and hence are marked hidden. Even this comment is not going to solve the issue in any way but it has to be made to clarify things.

**Comment by iammosespaulr** [comment]
@sylee957 what about printing multinomials ... like print_binomial in latex.py and mathml.py ?

**Comment by czgdp1807** [comment]
> Some people prefer to use
> 
> * ![image](https://user-images.githubusercontent.com/34944973/77272609-f6673380-6cf4-11ea-911e-70d0526f4101.png)

I think this is better than other options as it isn't redundant, won't require, `n = Sum(k1, k2, k3, k4, ..., kn)`.

**Comment by czgdp1807** [comment]
And binomial coefficients are multinomial coefficients. So, if they both are represented by classes then former should inherit the later.

**Comment by iammosespaulr** [comment]
> And binomial coefficients are multinomial coefficients. So, if they both are represented by classes then former should inherit the later.

That's what I asked @sylee957 ... but he was talking about writing one from scratch ... 
>In my opinion, I'd just implement it from scratch because expanding a multinomial into a product of binomials may not have a unique representation.
But the computation can be done internally by dispatching them to binomials.

**Comment by iammosespaulr** [comment]
> And binomial coefficients are multinomial coefficients. So, if they both are represented by classes then former should inherit the later.
But I guess this a better chance at generalisation ... doesn't it ?

**Comment by czgdp1807** [comment]
> Yes, but I'd expect it to be presented as a `class multinomial(Function):`

We can inherit `multinomial` in `binomial`. If `binomial` is already there then current APIs for `binomial` shouldn't break. If not then may be wen can go with the first option given by @sylee957 in the OP.

**Comment by iammosespaulr** [comment]
Okey 👍

**Comment by czgdp1807** [comment]
Though I am not sure which design scheme should be followed. I think that, having general classes subclassed in specialised classes help in getting things done with little code. Like, a truck is a vehicle so we should go for vehicle being subclassed in truck. Though if there is any other compelling reason to do the reverse then that should be presented here.

**Comment by sylee957** [comment]
When inheriting the class, the args structure must be coherent or other issues can arise.

I agree that multinomial without n looks better. However, it can make more difficult to keep the consistent structure with classes, so they should not inherit each other.

But I’d rather have other ideas like Binomial automatically evaluate to multinomial and keep binomial a stub class for notational purpose, if it should be

**Comment by czgdp1807** [comment]
I see, `binomial` is already there. May be then API for binomial can be extended to multinomial as well so that the consistency is maintained. It would be better if they both are compatible with each other.

**Comment by Arpan612** [comment]
So can we use the binomial class then?

**Comment by czgdp1807** [comment]
Can you please elaborate more on your question?

**Comment by Arpan612** [comment]
Could we implement the multinomial coefficient as a function or it is necessary to construct a separate class for it?

**Comment by czgdp1807** [comment]
I think that the APIs for `binomial` and `multinomial` should be similar with inter operability.

**Comment by Arpan612** [comment]
Sure sir.

**Comment by sylee957** [comment]
The only thing I'm concerned of making multinomial similar to binomial is that we will eventually see
![image](https://user-images.githubusercontent.com/34944973/77548923-89b98800-6ef2-11ea-97f8-4d61bc432326.png) and having to derive all the formula about it

**Comment by czgdp1807** [comment]
> The only thing I'm concerned of making multinomial similar to binomial is that we will eventually see
> ![image](https://user-images.githubusercontent.com/34944973/77548923-89b98800-6ef2-11ea-97f8-4d61bc432326.png) and having to derive all the formula about it

IMHO, this isn't a really good representation. It is redundant, self-correcting i.e., if k1, k2, k3, k4, ...kn sum less than `n` then the last term in the denominator will correct that mistake. So, I think, what we can do to ensure inter operability between the two is if we want to represent `binomial(n, k)` in terms of `multinomial` by `multinomial(k, n-k)`. Actually `nCk` is very popular as `binomial`(in fact I am not aware of better notations to represent binomial), however, as you have suggested multinomial has multiple representation and making it similar to `binomial` will lead to non-intuitive API.

**Comment by czgdp1807** [comment]
> I think that the APIs for `binomial` and `multinomial` should be similar with inter operability.

Just for clarification, by similarity I mean that functions provided by `multinomial` and `binomial` should be same. Making them exactly same would not be possible due to the reasons mentioned above. Please don't this point wrongly. Inter operation is what matters more IMO.

**Comment by oscarbenjamin** [comment]
> When inheriting the class, the args structure must be coherent or other issues can arise.

I think it is okay for subclasses to have different args. The important thing is that code should access properties rather than accessing the args directly. Then a subclass that changes the args structure just needs to override the properties that access the args.

**Comment by Arpan612** [comment]
@czgdp1807 @oscarbenjamin Could you please go through my PR?

**Comment by iammosespaulr** [comment]
```
>>> from sympy import *
>>> for i in range(10,15):
...     for j in range(10,15):
...             print(binomial(j,i), multinomial((i,j-i)))
... 
1 1
11 11
66 66
286 286
1001 1001
0 0
1 1
12 12
78 78
364 364
0 0
0 0
1 1
13 13
91 91
0 0
0 0
0 0
1 1
14 14
0 0
0 0
0 0
0 0
1 1
>>> multinomial(symbols('a b c'))
gamma(a + b + c + 1)/(gamma(a + 1)*gamma(b + 1)*gamma(c + 1))
>>> 
```
I've been working on this ... is this close to what you were expecting @sylee957 @czgdp1807 @oscarbenjamin

**Comment by czgdp1807** [comment]
IMO, it looks fine to me. Can you please make a PR to show whatever you have written for `multinomial`. I am not sure if multiple PRs are allowed but it is fine to make one, the better one can be kept.

**Comment by czgdp1807** [comment]
> IMO, it looks fine. Can you please make a PR to show whatever you have written for `multinomial`. I am not sure if multiple PRs are allowed but it is fine to make one, the better one can be kept.
@iammosespaulr

**Comment by iammosespaulr** [comment]
I'll submit a pr ... in 6-7 hrs or so. I'm soo sleepy , btw I am yet to add support for expanding the expression and for .eval , the rest are almost done @czgdp1807

**Comment by sylee957** [comment]
The framework I'd begin with is 

```python3
from sympy import *
from sympy.core.sympify import _sympify
from sympy.functions.combinatorial.factorials import CombinatorialFunction

class multinomial(CombinatorialFunction):
    @classmethod
    def eval(cls, *k):
        k = [_sympify(x) for x in k]
        n = Add(*k)
        if all(x.is_Integer for x in k):
            n = Add(*k)
            numer = factorial(n)
            denom = Mul(*(factorial(x) for x in k))
            return numer/denom
        
    def _eval_rewrite_as_factorial(self, *args, **kwargs):
        n = Add(*args)
        numer = factorial(n)
        denom = Mul(*(factorial(x) for x in args))
        return numer/denom
    
    def _eval_rewrite_as_binomial(self, *args, **kwargs):
        pass
    
    def _eval_rewrite_as_gamma(self, *args, **kwargs):
        pass
```
Actually, when following the sympy `Function` framework, I don't think that it's easy to access parameters via properties because it uses classmethod to evaluate. And other things like sorting the `k` is not easy with this framework.

**Comment by iammosespaulr** [comment]
@sylee957 I wrote something similar to that ... but It was just using gamma

**Comment by sylee957** [comment]
For integers only, it’s better to use factorials

**Comment by iammosespaulr** [comment]
Yes!

**Comment by iammosespaulr** [comment]
@sylee957 @czgdp1807 I've drafted a PR ... lmk what you think.

**Comment by sylee957** [comment]
From the wolfram's source, mutinomial is undefined when `n` is negative integer (Because the gamma and factorial gets division of infinity by infinity) So when this is encountered, it should raise an error.

![image](https://user-images.githubusercontent.com/34944973/77647002-ff802b00-6fa8-11ea-93cb-6e22e57205f6.png)

**Comment by iammosespaulr** [comment]
Yeahh. I've removed test cases where n is negative. I'll add a check too.

**Comment by Arpan612** [comment]
@sylee957 Okay sir.

**Comment by Arpan612** [comment]
@sylee957 @czgdp1807 Sir, I believe my version of the PR is slightly better. Also, I have squashed the commits for easy readability. Due to the squash, checks are been performed all over again. I request you to wait for some time.

**Comment by Arpan612** [comment]
@sylee957 @czgdp1807 Please review and merge PR #18952

**Comment by smichr** [comment]
#18906 had code to calculate a given multinomial coefficient but it is stalled. It would be nice to have the ability to calculate a concrete coefficient. The following is an efficient version that could be used:
```python
def multinomial_coefficient(*e):
    """Return the multinomial coefficient (e1 + e2 + ...)! / e1! / e2! / ..."""
    # https://stackoverflow.com/a/56856279/1089161
    if not e:  # no parameters
        return 1
    ibig = e.index(max(e))
    num = e[ibig] + 1
    rv = 1
    for n in e[:ibig] + e[ibig + 1:]:
        for den in range(1, n + 1):
            rv *= num // den
            num += 1
    return rv
```
Note that currently one can calculate *all* coefficients with `multinomial_coefficients` but if you only want a few (or one) then something like the above would be useful.
